tracer.go

```diff
@@ -1,48 +1,72 @@
 package zipkin
 
 import (
+	"context"
+	"sync/atomic"
 	"time"
 
-	"github.com/openzipkin/zipkin-go/kind"
+	"github.com/openzipkin/zipkin-go/idgenerator"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/propagation"
+	"github.com/openzipkin/zipkin-go/transport"
 )
 
 // Tracer is our Zipkin tracer implementation.
 type Tracer struct {
-	options TracerOptions
+	noop                 int32
+	localEndpoint        *model.Endpoint
+	sharedSpans          bool
+	sampler              Sampler
+	generate             idgenerator.IDGenerator
+	defaultTags          map[string]string
+	unsampledNoop        bool
+	extractFailurePolicy ExtractFailurePolicy
+	transport            transport.Transporter
 }
 
 // NewTracer returns a new Zipkin Tracer.
-func NewTracer(options ...TracerOption) (*Tracer, error) {
+func NewTracer(transport transport.Transporter, options ...TracerOption) (*Tracer, error) {
 	// set default tracer options
-	opts := &TracerOptions{
+	t := &Tracer{
 		sharedSpans: true,
 		sampler:     alwaysSample,
-		generate:    &RandomID64{},
+		generate:    idgenerator.NewRandom64(),
 		defaultTags: make(map[string]string),
-		transport:   &NoopTransport{},
+		transport:   transport,
 	}
 
 	// process functional options
 	for _, option := range options {
-		if err := option(opts); err != nil {
+		if err := option(t); err != nil {
 			return nil, err
 		}
 	}
 
-	return &Tracer{options: *opts}, nil
+	return t, nil
 }
 
-// StartSpan creates and starts a span
-func (t *Tracer) StartSpan(
-	name string, kind kind.Type, options ...SpanOption,
-) Span {
+// StartSpanFromContext creates and starts a span using the span found in
+// context as parent. If no parent span is found a root span is created.
+func (t *Tracer) StartSpanFromContext(ctx context.Context, name string, options ...SpanOption) (Span, context.Context) {
```

Comment:
fyi I use nextSpan (just as it disambiguates nextChild from maybe new trace). you don't have to of course.

Reply:
It needs to signal more than just nextSpan. It also signals Go's Context to be in play as it not only returns a new (child) Span but also returns a new Go context with the created Span inside it. Naming is same as with OT so no confusion will happen there either.

Reply:
This sounds very OT, in addition you could create the span without starting it. What about `nextSpan` passing `context.Context`? It makes sense as `nextSpan` considers by default the current span context which in golang is injected in the context. For noop you could use `context.Background()`. I don't think there is such a case where you don't pass a context.

Reply:
I personally like being explicit about Go's Context. In local tracing it is very well possible to not use Go's context and with the separation of StartSpan and SpanFromContext I can make that conscious decision to use or not use Context.